### PR TITLE
[inductor][cpp][gemm] improve large bs perf with better cache blocking

### PR DESCRIPTION
**Pull request by jgong5**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133538
* #133447
* #131306
* #133312
* #132730
* __->__ #132729

Improve the cache blocking by reducing Mc_blocks to make A reside in L2 and reused by B as much as possible. This improves large bs perf for both scenarios: 1) N is large and K is of medium sizes; 2) K is large. Different strategies are used to handle these scenarios. Check the notes in `get_cache_blocking` in the changes.

Measured with 56-core Intel (R) Xeon (R) CPU Max 9480, jemalloc 5.1 and intel omp, bf16. Run with code cache of B matrix (weights).

Model Shapes | Before Optimization | After Optimization | Speedup | onednn linear | Speedup over onednn
-- | -- | -- | -- | -- | --
M=1024, N=12288, K=4096 (Llama2-8b) | 5.69 ms | 3.71 ms | 1.53 | 4.53 ms | 1.22
M=1024, N=4096, K=4096 (Llama2-8b) | 1.69 ms | 1.63 ms | 1.04 | 2.05 ms | 1.26
M=1024, N=22016, K=4096 (Llama2-8b) | 10.32 ms | 6.57 ms | 1.57 | 8.46 ms | 1.29
M=1024, N=4096, K=11008 (Llama2-8b) | 5.21 ms | 3.26 ms | 1.60 | 4.65 ms | 1.43
M=1024, N=5120, K=4096 (Llama3-8b) | 1.99 ms | 1.78 ms | 1.12 | 2.31 ms | 1.30
M=1024, N=28672, K=4096 (Llama3-8b) | 13.41 ms | 8.56 ms | 1.57 | 10.96 ms | 1.28
M=1024, N=4096, K=14336 (Llama3-8b) | 6.93 ms | 4.31 ms | 1.61 | 6.24 ms | 1.45

cc @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by leslie-fang-intel — APPROVED** [review]

**Review by chunyuan-w — APPROVED** [review]

File: torch/_inductor/config.py [hunk 1]
```diff
@@ -741,6 +741,12 @@ class cpp:
     # When set to 0, the number of slices is unlimited.
     gemm_max_k_slices = int(os.environ.get("TORCHINDUCTOR_CPP_GEMM_MAX_K_SLICES", "1"))
 
+    # For perf tuning and debugging purpose, configure the pre-defined cache blocking for
+    # MxNxK dims respectively. The blockings are separated by comma and the unit is
+    # the number of register blocks.
+    # For example, "4,1,10" means 4 register blocks on M, 1 on N and 10 on K respectively.
+    gemm_cache_blocking = os.environ.get("TORCHINDUCTOR_CPP_GEMM_CACHE_BLOCKING", None)
```
Comment (jansel) [inline torch/_inductor/config.py:748]: Please add a test that enables this config via `@config.patch(gemm_cache_blocking="...")`

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/config.py (hunk 1 above)
Comment (jgong5) [reply torch/_inductor/config.py:751]: Thanks. Added.

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 2]
```diff
@@ -365,33 +377,62 @@ def get_cache_blocking(register_blocking, thread_blocking):
...
+                # A (Mc x Kc) in L2 by B (Kc x Nc). C (Mc x Nc) resides in L2.
+                Mc_blocks = Mt_blocks
+                Nc_blocks = min(math.ceil(Mc_blocks * M0 / N0), Nt_blocks)
+                Nc_bytes = Nc_blocks * N0 * 4  # assume C or acc is float32/int32
+                Kc_bytes = Kc_blocks * K0 * num_byte_A
+                if Mc_blocks * M0 * (Kc_bytes + Nc_bytes) > L2:
+                    # The following is the solution for 4*Mc*Nc + Mc*Kc_bytes = L2,
+                    # assuming Mc == Nc for good data reuse.
+                    M_max = (math.sqrt(Kc_bytes * Kc_bytes + 16 * L2) - Kc_bytes) / 8
+                    if M_max < Mc_blocks * M0:
```
Comment (leslie-fang-intel) [inline torch/_inductor/codegen/cpp_gemm_template.py:433]: Should it be an `assert` theoretically? But considering we use some approximated calculation. I guess it should be fine.

**Review by leslie-fang-intel — APPROVED** [review]

File: torch/_inductor/codegen/cpp_gemm_template.py (hunk 2 above)
Comment (jgong5) [reply torch/_inductor/codegen/cpp_gemm_template.py:434]: It is not an assertion. When it is false, we use the default one which is `Mt_blocks`.

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 3]
```diff
@@ -124,28 +126,32 @@
...
-            {{ kernel.define_buffer(acc_buf_name, ["m_end - m_start", "N0"], acc_buf_dtype) }}
+            {{ kernel.define_buffer(acc_buf_name, ["m_end - m_start", "Nc_blocks*N0"], acc_buf_dtype) }}
             {%- endif %}
-            for (int64_t nc = n_block_start; nc < n_block_end; ++nc) {
+            for (int64_t nc = n_block_start; nc < n_block_end; nc += Nc_blocks) {
                 const int64_t n_start = nc * N0;
-                const int64_t n_end = std::min((nc + 1) * N0, N);
+                const int64_t n_end = std::min(std::min(nc + Nc_blocks, n_block_end) * N0, N);
                 const int64_t n_size = n_end - n_start;
+                const int64_t nc_block_end = std::min(nc + Nc_blocks, n_block_end); // FIXME: maybe exceeding N?
```
[the remark this answers is not in the document]
Comment (jgong5) [reply torch/_inductor/codegen/cpp_gemm_template.py]: Good point. Updated the comment.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**